### PR TITLE
Pin napari and pydantic when installing a plugin

### LINKED ISSUE #5279: BUG: Plugins can downgrade napari via plugin GUI, without warning the user
## 🐛 Bug

If a plugin pins napari to an older version than the running napari with plugin installer GUI, the plugin will install and downgrade napari, silently without warning to the user.

## To Reproduce

1. install napari 0.4.16 or 0.4.17rc
2. use the plugin GUI to install `napari-n2v`

## Expected behavior

napari should resist version changes initiated by a plugin and should instead raise a warning to the user that such and such plugin is trying to downgrade napari and that to force install it—which may break your environment—please use the command line or make a fresh env.

## Environment

fresh conda-forge install of napari 0.4.16, python 3.9, macOS 12.6, arm64

### PR DESCRIPTION
**Pull request by Czaki**
# Fixes/Closes

Closes #5279

# Description

Backport changes from new plugin dialog to 0.4.18 release 

# References

## Type of change

- [x] Enhancement (non-breaking improvements of an existing feature)

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change
- [ ] example: I check if my changes works with both PySide and PyQt backends
      as there are small differences between the two Qt bindings.  

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

### CODE REVIEW

**Comment by Czaki** [comment]
@goanpeca Did the assumption (present currentyl in the code) that the installer could be only pip or mamba?

**Review by jni — APPROVED** [review]
@goanpeca @jaimergp I'd love your input on whether this will work for napari 0.4.18 on conda-forge — it is the last PR holding up 0.4.18! 🥳

**Comment by jni** [comment]
Well, I'm going to merge and assume this is ok as I'd like to get the final RC out today. Since afaik this fixes a problem that existed in many prior versions, I don't think it's a *huge* deal if it doesn't actually work. 😅

File: napari/_qt/dialogs/qt_plugin_dialog.py [hunk 1]
```diff
@@ -58,6 +61,23 @@
 InstallerTypes = Literal['pip', 'mamba']
 
 
+def _pip_constraints():
+    return [f"napari=={__version__}", "pydantic<2"]
+
+
+def _conda_constraints():
+    return [f"napari={__version__}", "pydantic<2.0a0"]
```
Comment (jaimergp) [inline napari/_qt/dialogs/qt_plugin_dialog.py:69]: This is better handled in a repodata patch (pip doesn't have this mechanism but conda-forge does). Now this constraint will be tied to that release forever, which is maybe not a problem in this case, but we need to be careful about internalizing "external" problems in our codebase.

Comment (Czaki) [reply napari/_qt/dialogs/qt_plugin_dialog.py:69]: Could you point where? Also this release is not compatibility with pydantic 2 and never will be.

Comment (jni) [reply napari/_qt/dialogs/qt_plugin_dialog.py:69]: Yeah @jaimergp note that this change is on a branch that won't live beyond this release. So nothing is internalised, it is forgotten tomorrow. 😂

Comment (jni) [reply napari/_qt/dialogs/qt_plugin_dialog.py:69]: Looks like `_conda_constraints` is not used anywhere anyway, so I'll delete it in the [now-necessary follow-up PR](https://forum.image.sc/t/request-for-testing-napari-0-4-18-release-candidate/82833/21).

**Pull request merged**